tools/src/test/java/org/apache/kafka/tools/PushHttpMetricsReporterTest.java

```diff
@@ -0,0 +1,333 @@
+                new Object(),
+                new MetricName("name2", "group2", "desc2", Collections.singletonMap("key2", "value2")),
+                new ImmutableValue(2.0),
+                null,
+                time
+        );
+        KafkaMetric metric3 = new KafkaMetric(
+                new Object(),
+                new MetricName("name3", "group3", "desc3", Collections.singletonMap("key3", "value3")),
+                new ImmutableValue(3.0),
+                null,
+                time
+        );
+        reporter.init(Arrays.asList(metric1, metric2));
+        reporter.metricChange(newMetric1); // added in init, modified
+        reporter.metricChange(metric3); // added by change
+        reporter.metricRemoval(metric2); // added in init, deleted by removal
+
+        reportRunnable.getValue().run();
+        JsonNode payload = new ObjectMapper().readTree(httpPayload.getValue());
+        assertTrue(payload.isObject());
+        assertPayloadHasClientInfo(payload);
+
+        // We should be left with the modified version of metric1 and metric3
+        JsonNode metrics = payload.get("metrics");
+        assertTrue(metrics.isArray());
+        assertEquals(2, metrics.size());
+
+        JsonNode m1 = metrics.get(0);
+        assertEquals("name1", m1.get("name").textValue());
+        assertEquals("group1", m1.get("group").textValue());
+        JsonNode m1Tags = m1.get("tags");
+        assertTrue(m1Tags.isObject());
+        assertEquals(1, m1Tags.size());
+        assertEquals("value1", m1Tags.get("key1").textValue());
+        assertEquals(-1.0, m1.get("value").doubleValue(), 0.0);
+
+        JsonNode m3 = metrics.get(1);
+        assertEquals("name3", m3.get("name").textValue());
+        assertEquals("group3", m3.get("group").textValue());
+        JsonNode m3Tags = m3.get("tags");
+        assertTrue(m3Tags.isObject());
+        assertEquals(1, m3Tags.size());
+        assertEquals("value3", m3Tags.get("key3").textValue());
+        assertEquals(3.0, m3.get("value").doubleValue(), 0.0);
+
+        reporter.close();
+
+        verifyAll();
+    }
+
+    private void expectConfigure() {
+        EasyMock.expect(
+                executor.scheduleAtFixedRate(EasyMock.capture(reportRunnable), EasyMock.eq(5L), EasyMock.eq(5L), EasyMock.eq(TimeUnit.SECONDS))
+        ).andReturn(null); // return value not expected to be used
+    }
+
+    private void configure() {
+        Map<String, String> config = new HashMap<>();
+        config.put(PushHttpMetricsReporter.METRICS_URL_CONFIG, URL.toString());
+        config.put(PushHttpMetricsReporter.METRICS_PERIOD_CONFIG, "5");
+        reporter.configure(config);
+    }
+
+    private void expectRequest(int returnStatus) throws Exception {
+        expectRequest(returnStatus, false);
+    }
+
+    // Expect that a request is made with the given response code
+    private void expectRequest(int returnStatus, boolean readResponse) throws Exception {
```

Comment:
Oh my, that is quite some mock code!

Reply:
I did consider a non-strict mock here since there are so many calls. Some are useful to verify (e.g. content type), others seem less so (e.g. `setUseCaches`. I'm happy to switch to a non-strict mock to cut down on test noise if that's preferred.